Kill allocations in GraphicsPath/Graphics

JeremyKuhne opened the pull request:
This removes a great deal of the unneeded allocations in GraphicsPath and Graphics. Points and rectangles match the native types exactly and don't need special marshalling outside of pinning arrays.

This removes all usages of GPPOINT/F.  Many GPRECT/F usages were removed, but still exist in other classes.

The first commit is #30237, which is also in PR

JeremyKuhne commented:
My intent was to look at overlap with the Unix files in a follow up. Didn't realize that @hughbe had a pending PR here to do some of that. (Sorry @hughbe)

JeremyKuhne commented:
These changes make a big impact on memory usage and performance for Graphics/GraphicsPath. There is a lot more to do in System.Drawing in this same vein. As mentioned there is GPRECT/F to be finished, but there is plenty more of other unnecessary marshalling contortions.

danmosemsft commented:
When merging please don't squash formatting and functional commits obviously

JeremyKuhne commented:
> When merging please don't squash formatting and functional commits obviously

Yep, I was keeping them as separate PRs as well (see #30237).

Tanya-Solyanik commented on src/System.Drawing.Common/src/System/Drawing/Drawing2D/PathGradientBrush.cs:
```diff
@@ -13,47 +13,36 @@ public sealed class PathGradientBrush : Brush
-                {
-                    Marshal.FreeHGlobal(pointsBuf);
-                }
+                SafeNativeMethods.Gdip.CheckStatus(SafeNativeMethods.Gdip.GdipCreatePathGradient(
```
Consider adding "using SafeNativeMethods". BTW, could you please point me to the .net core style guide?

danmosemsft replied:
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

JeremyKuhne replied:
> Consider adding "using SafeNativeMethods"

Yeah, I intend to do that in a separate change. Discussed with @safern offline. :)

Tanya-Solyanik commented on src/System.Drawing.Common/src/System/Drawing/Graphics.Unix.cs:
```diff
@@ -1046,7 +1039,7 @@ public void DrawPath(Pen pen, GraphicsPath path)
                 throw new ArgumentNullException("pen");
             if (path == null)
                 throw new ArgumentNullException("path");
```
nameof? unless it is slotted for the next PR.

danmosemsft replied:
this may help - https://github.com/dotnet/corefx/pull/30228#issuecomment-395936113

JeremyKuhne replied:
I wasn't updating the Unix files in this PR as some of them will fold into shared code.

Tanya-Solyanik commented on src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs:
```diff
@@ -417,85 +355,62 @@ public InterpolationMode InterpolationMode
-                SafeNativeMethods.Gdip.CheckStatus(status);
-
-                return dpi[0];
+                SafeNativeMethods.Gdip.CheckStatus(SafeNativeMethods.Gdip.GdipGetDpiX(new HandleRef(this, NativeGraphics), out float dpi));
```
Long code lines don't wrap well in the default github editor, would you mind breaking them up in one of theupcoming PRs?

danmosemsft replied:
@Tanya-Solyanik you might try https://userstyles.org/styles/108591/github-wide. It is great for reading code/fixed format text on github.

danmosemsft commented on src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs:
```diff
@@ -776,17 +667,14 @@ public void DrawArc(Pen pen, Rectangle rect, float startAngle, float sweepAngle)
-            int status = SafeNativeMethods.Gdip.GdipDrawBezier(new HandleRef(this, NativeGraphics), new HandleRef(pen, pen.NativePen), x1, y1,
-                                                x2, y2, x3, y3, x4, y4);
-
-            //check error status sensitive to TS problems
```
@Tanya-Solyanik you might try https://userstyles.org/styles/108591/github-wide. It is great for reading code/fixed format text on github.

JeremyKuhne replied:
The status helper discusses what it is doing. I plan to revamp the error checking helpers to make it a bit more understandable as to what helper is doing what.

Tanya-Solyanik commented on src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs:
```diff
@@ -2217,103 +1865,77 @@ public void DrawImageUnscaledAndClipped(Image image, Rectangle rect)
-         *
-         *  @notes Perspective blt only works for bitmap images.
-         */
+        // Affine or perspective blt
```
Consider using `///` -style comments and appropriate tags

JeremyKuhne replied:
This one was a little weird. It was a general comment that applied to a number of methods.

JeremyKuhne commented on src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs:
```diff
@@ -1790,29 +1467,21 @@ public void FillClosedCurve(Brush brush, Point[] points, FillMode fillmode)
+                CheckErrorStatus(SafeNativeMethods.Gdip.GdipFillClosedCurve2I(
+                    new HandleRef(this, NativeGraphics),
+                    new HandleRef(brush, brush.NativeBrush),
+                    p, points.Length,
```
It seemed a little crazy to have single line separation for things like x, y, z.  I put related ones together when I had to break up the full line. The same was done in other places.

JeremyKuhne replied:
> The same was done in other places.

Prior to this change, that is.

safern commented:
I think we should just keep the https://github.com/dotnet/corefx/pull/30244/commits/c31580a8da26de9696e3692ffce7b0e3e9c5f603 and then in @hughbe's https://github.com/dotnet/corefx/pull/29713 the cleanup commit is addressed.

JeremyKuhne commented:
Rebased on @hughbe changes. Also added a commit for `using Gdip = System.Drawing.SafeNativeMethods.Gdip;` per discussions with @safern.

Hope to kill the rest of GPRECT/F usages soon.

JeremyKuhne commented:
@dotnet-bot test Linux arm Release Build
@dotnet-bot test Tizen armel Debug Build

ubuntu1404-20170120 seems hung up?
